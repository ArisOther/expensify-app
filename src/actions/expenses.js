import uuid from 'uuid'

//======================== ACTION GENERATOR ==========================
//-- 2. Berisi Nilai Defaut State dan deklarasi STATE, state ini yang akan diolah REDUCER
    // ADD_EXPENSE 
    export const addExpense = (
        {
            //Ini adalah argumen dalam function
            //Memberi default value pada STATE,
            //Nilai ini yang akan diambil, jika tidak ada perubahan nilai di DISPATCH
            description='',
            note='', 
            amount=0, 
            createdAt=0,
            //Date='53' 
        } = {}
    ) => ({
        //Deklarasi STATE
        //property2 STATE ini bisa diolah/diubah nilainya di dispatch action,
        //Nilai akhir setelah perubahan di DISPATCH akan diolah oleh REDUCER
        type: 'ADD_EXPENSE', //sebagai ID untuk REDUCER
        expense: {
            id: uuid(),
            description,
            note,
            amount,
            createdAt,
            //Date
        }
    })


    //---------------------- add_expense end -------------------

    // REMOVE_EXPENSE
    export const removeExpense = ({id} = {}) => ({
        type: 'REMOVE_EXPENSE',
        id
    })

    //EDIT_EXPENSE
    export const editExpense = (id, updates) => ({
        type : 'EDIT_EXPENSE',
        id,
        updates

    })