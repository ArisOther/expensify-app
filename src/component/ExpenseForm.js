import React from 'react'
import moment from 'moment'
import {SingleDatePicker} from 'react-dates'
import 'react-dates/lib/css/_datepicker.css'
import 'react-dates/initialize'

const now = moment()
console.log(now.format('MMM Do, YYYY'))
//class base component
export default class ExpenseForm extends React.Component {
    //state
    //dirubah bentuk constructor supaya bisa diolah editExpense.js
    constructor(props){
        super(props)
        this.state = {
            description: props.expense ? props.expense.description:'',
            note:props.expense ? props.expense.note:'',
            amount:props.expense ? (props.expense.amount/100).toString():'',
            createdAt: props.expense ? moment(props.expense.createdAt):moment(),
            calendarFocused:false,
            error:''
        }
    }
    
    onDescriptionChange=(e) => {
        const description = e.target.value //state description = mendapatkan input dari form
        this.setState(()=>({description})) //merubah state description menjadi nilai diatas
    }
    onNoteChange=(e)=>{
        const note = e.target.value
        this.setState(()=>({note}))
    }
    onAmountChange=(e)=>{
        const amount = e.target.value
        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){//regex-format 2 angka dibelakang koma - regex101.com 
            this.setState(()=>({amount}))
        }
        
    }
    onDateChange = (createdAt) => {
        if (createdAt) {
            this.setState(()=>({createdAt}))
        }
    }
    onFocusChange = ({focused})=>{
        this.setState(()=>({calendarFocused: focused}))
    }
    onSubmit = (e)=>{
        e.preventDefault()//mematikan fungsi bawaan
    
        if(!this.state.description || !this.state.amount){//cek jika desc / amount kosong
            this.setState(()=>({error:'ini ada masalah'}))
        } else {
            this.setState(()=>({error:''}))//jika form terisi, pesan error kosong
            this.props.onSubmit({ //bentuk props karena onSubmit akan dipanggil sebagai child di AddExpense
                description: this.state.description,
                amount: parseFloat (this.state.amount,10)*100, //memparsing bilangan bulat
                createdAt:this.state.createdAt.valueOf(), //valueOf = member moment.js yang berfungsi mendapatkan waktu ms
                note:this.state.note
            })
        }
        
    }     
    
    render() {
        return (
            <div>
                {/* Mencetak state.error jika ada isinya */}
                {this.state.error && <p>{this.state.error}</p>}

                <form onSubmit={this.onSubmit}>
                    <input 
                    type="text" 
                    placeholder="Description"
                    autoFocus
                    value={this.state.description}
                    onChange={this.onDescriptionChange}
                    /> 
                    
                    <input
                    type="text"
                    placeholder="amount"
                    value={this.state.amount}
                    onChange={this.onAmountChange}
                    />

                    {/* component dari react-dates */}
                    <SingleDatePicker
                    date={this.state.createdAt}
                    onDateChange={this.onDateChange}
                    focused={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={()=>false}
                    // startDateId={''}
                    // endDateId={''}
                    />

                    <textarea
                    type="text"
                     placeholder="Add a note for your expense (optional)"
                     value={this.state.note}
                     onChange={this.onNoteChange}
                    >

                    </textarea>
                    <button>Add Expense</button>
                    
                </form>
            </div>
        )
    }
}