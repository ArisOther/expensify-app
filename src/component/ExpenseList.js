import React from 'react'
import {connect} from 'react-redux'
import ExpenseListItem from './ExpenseListItem'
import selectExpenses from '../selectors/expenses'

//bentuk props
//komponen ini yang di tampilkan/dipanggil oleh ExpenseDashboard, oleh karena itu bentuknya props
const ExpenseList = (props) => (
    <div>
        <h1>Expense List</h1>
        {/* Mengambil data -expenses, filters- dari bawah (mapStateProps) */}
        {props.expenses.map((expense) =>{// expenses = state awal, expense = nama objek yang menunjukan per individu
            return <ExpenseListItem key={expense.id} {...expense}/>
        })}
        {props.filters.text}
    </div>
)

//state
//sumber data state dari ExpenseList diatas
const mapStateToProps = (state) => {
    return {
        //objek baru/props, mengambil nilai dari state action
        expenses: selectExpenses(state.expenses, state.filters),
        filters: state.filters
    }
}

//menghubungkan mapStateToProps dengan ExpenseList
export default connect(mapStateToProps)(ExpenseList)