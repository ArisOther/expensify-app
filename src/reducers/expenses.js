//========================== REDUCER ==========================================
//== 2. reducer mengolah STATE dan logic mengambil dari ACTION DISPATCH & ACTION GENERATOR
        //============= Expenses Reducer===============================
        const expensesReducerDefaultState = []

        //const expensesReducer = (state = expensesReducerDefaultState, action) => {
        export default (state = expensesReducerDefaultState, action) => {
            // mendapatkan data action.type dari ACTION DISPATCH
            switch (action.type){
                case 'ADD_EXPENSE':
                    //return state.concat(action.expense)
                    return [
                        //state dari ADD_EXPENSE bentuknya array
                        // ...state --> spread operator, mengeluarkan data dari array sehingga berdiri masing2
                        //action.expense-->dari ACTION GENERATOR
                        ...state, 
                        action.expense  
                    ]
                case 'REMOVE_EXPENSE' : 
                    return state.filter(({id}) => {
                        return id !== action.id //true
                    })
                case 'EDIT_EXPENSE' :
                    return state.map((expense) => {
                        if (expense.id === action.id){
                            return {
                                ...expense, //men-spread / menampilkan semua isi array
                                ...action.updates //menampilkan hasil perubahan
                            }
                        } else {
                            return expense
                        }
                    })

                default:
                    return state
            }

        }
        //------------- expenses reducer end-----------------------------