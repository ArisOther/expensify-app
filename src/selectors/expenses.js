//pada app.js func ini dipanggil GetVisibleExpenses
//meskipun sebenarnya nama bisa diganti apa saja karena struktur export
export default (expenses, {text, sortBy, startDate, endDate}) => {
    return expenses.filter((expense) => {
        //logic untuk cek kondisi filter
        //membandingkan  startdate milik Filters dengan createdAt milik expenses
        const startDateMatch = typeof startDate !=='number' || expense.createdAt >= startDate
        const endDateMatch = typeof endDate !=='number' || expense.createdAt <= endDate
        //menterjemahkan input pencarian yang UPPERCASE menjadi LOWERCASE
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase())

        return startDateMatch && endDateMatch && textMatch
        //mengurutkan hasil filter 'apa saja'  berdasarkan Nilai tertinggi
    }).sort((a,b) => {
        if(sortBy === 'date'){
            return a.createdAt < b.createdAt ? 1:-1
        } else if (sortBy === 'amount'){
            return a.amount < b.amount ? 1:-1
        }
    })
}