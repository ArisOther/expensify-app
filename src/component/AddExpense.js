import React from 'react'
import {connect} from 'react-redux'
import ExpenseForm from './ExpenseForm'
import {addExpense} from '../actions/expenses'

const AddExpensePage = (props)=> (
    <div>
        <h1>Add Expense</h1>
        <ExpenseForm //memanggil komponen form
        onSubmit={(expense)=>{
            //dimatikan karena -addExpense(expense)- tidak ramah testing, penggantinya adalah props.onSubmit dibawah
            // props.dispatch(addExpense(expense)) //dispatch berdasarkan data input dan mengembalikan ke seluruh state expense
            
            props.onSubmit(expense)
            props.history.push('/') //setelah submit, redirect ke halaman depan
        }
    }
        />
    </div>
)
//Demi test - pembuatan mapDispatchToProps untuk menjembatani onSubmit dengan proses dispatch
const mapDispatchToProps = (dispatch)=>({
    onSubmit: (expense) => dispatch(addExpense(expense))
})

export default connect(undefined, mapDispatchToProps )(AddExpensePage)