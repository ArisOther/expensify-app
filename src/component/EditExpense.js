import React from 'react'
import {connect} from 'react-redux'
import ExpenseForm from './ExpenseForm'
import {editExpense, removeExpense} from '../actions/expenses'

const EditExpensePage = (props) => {
    
    return (
        <div>
            <ExpenseForm
                expense={props.expense}
                onSubmit={(expense)=>{
                    props.dispatch(editExpense(props.expense.id, expense)) //akan merubah item berdasarkan id
                    props.history.push('/') //setelah tekan tombol edit, akan redirect ke hal. utama
                    
                }}
                />

            <button onClick={() => { //onClick Remove button
                props.dispatch(removeExpense({id: props.expense.id}))//dispatch berdasarkan reducer removeExpense by 'id'
                props.history.push('/')
            }}>Remove</button>
             {/* <p>This is edit page with id of {props.match.params.id}</p>  */}
             {/* <p>This is edit page</p> */}
        </div>
    )
}

//const mapStateToProps = (state,props)
    //mengakses state karena akan dilakukan pencarian state - expenses array
    //apa yang di cari adalah props.match.params.id dimana itu adalah suatu props - data id dalam expenses array
const mapStateToProps = (state, props) => {
    return {
    //expense: ---> hanya nama object
    //state.expenses.find(()=>{ ---> mengakses state.expenses dan melakukan pencarian didalam array itu
    //expense.id ---> ini adalah id individual / per item (yang awalnya diolah di ExpenseList.js
        //(kenapa kok muncul expense sebagai data per item, prosesnya ada di ExpenseList.js)
    //=== props.match.params.id ---> cek expense.id yang sama dengan props.match.params.id
    expense: state.expenses.find((expense)=> expense.id === props.match.params.id)
}
}

 //connect()(EditExpensePage)---> memnungkinkan props.match.params.id diolah di mapStateToProps
 //connect()(mapStateToProps)(EditExpensePage)---> memnungkinkan keduanya terkonek
export default connect(mapStateToProps)(EditExpensePage)