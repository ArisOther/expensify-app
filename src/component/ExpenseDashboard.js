import React from 'react'
import ExpensesSummary from './ExpensesSummary'
import ExpenseListFilters from './ExpenseListFilters'
import ExpenseList from './ExpenseList'



const ExpenseDashboardPage = ()=> {
    return(
    <div>
        <ExpensesSummary/>
        <ExpenseListFilters/>
        <ExpenseList/>
    </div>
    )
}

export default ExpenseDashboardPage