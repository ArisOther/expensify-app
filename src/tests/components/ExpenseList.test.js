import React from 'react'
import {shallow} from 'enzyme'
import {ExpenseList} from '../../component/ExpenseList'
import expenses from '../fixtures/expenses'

test('Harus bisa merender ExpenseList menggunakan data dari fixtures/expenses',
() => {
    const wrapper = shallow(<ExpenseList expenses={expenses}/>)
    expect(wrapper).toMatchSnapshot()
}
)