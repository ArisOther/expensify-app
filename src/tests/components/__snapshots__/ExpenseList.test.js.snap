// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Harus bisa merender ExpenseList menggunakan data dari fixtures/expenses 1`] = `
<div>
  <h1>
    Expense List
  </h1>
  <ExpenseListItem
    amount={195}
    createdAt={0}
    description="gum"
    id="1"
    key="1"
    note=""
  />
  <ExpenseListItem
    amount={109500}
    createdAt={-345600000}
    description="rent"
    id="2"
    key="2"
    note=""
  />
  <ExpenseListItem
    amount={4500}
    createdAt={345600000}
    description="Credit Card"
    id="3"
    key="3"
    note=""
  />
</div>
`;
